libcontainer/container_linux.go

```diff
@@ -1545,6 +1556,59 @@ func (c *Container) Restore(process *Process, criuOpts *CriuOpts) error {
 	return err
 }
 
+// logCriuErrors tries to find and log errors from a criu log file.
+// The output is similar to what "grep -B5 Error" does.
+func logCriuErrors(dir, file string) {
+	lookFor := []byte("Error") // Print the line that contains this...
+	const before = 5           // ... and a few preceding lines.
+
+	logFile := filepath.Join(dir, file)
+	f, err := os.Open(logFile)
+	if err != nil {
+		logrus.Warn(err)
+		return
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	var lines [][]byte
+
+	lineNo := 0
+	lastLine := -1
+	didHeader := false
+	for s.Scan() {
+		lineNo++
+		line := s.Bytes()
+		lines = append(lines, line)
+		if len(lines)-1 > before {
+			// Remove the oldest line.
+			lines = lines[1:]
```

Comment:
nit: removing the first element from an array always looks suspicious. You can think to rework it like this:
lines[idx % before] = line;

Reply:
To me this looks like a legitimate trick when using Go slices. I mostly aimed for readability (and I have to admit it could be better).

Reply:
@avagin I redid this using a real circular buffer this time. Still looks a tad complicated but no slice tricks this time. PTAL.